README.md
# Uptime check
[![npm version](https://badge.fury.io/js/uptime-check.svg)](https://badge.fury.io/js/uptime-check) [![Coverage Status](https://coveralls.io/repos/github/sznowicki/uptime-check/badge.svg?branch=master)](https://coveralls.io/github/sznowicki/uptime-check?branch=master)

Checks uptime status of given url and provided keyword (optional).

## How it works
Makes simple curl requests, gathers data on response and if requested also is looking for a keyword inside the document body.

If http code (response) is 2xx website is considered "up". If there is a keyword given for search, but the keyword cannot be found in the document body, website is considered "down".

## v4 migration
From version 4 there is no `userAgent` option any more. If you used this, please migrate to:

```js
check({
  headers: {
      'User-Agent': 'Your custom UA string',
  }
});
```

## Example usage
```javascript
check({
  url: 'https://example.com',
  keyword: 'example',
  redirectsLimit: 0
})
  .then(result => {
    /*
        see index.d.ts:Result
    */
  })
  .catch((err) => {
    /*
    err = '{Error} unexpected error.';
    */
  });
```


CHANGELOG.MD
# Changelog

## v4.2.0
- Fixed type error when response.body is undefined
- Dependency update (got)
-
## v4.1.0
### Dependency removed
- Removed `header-parse` unused dependency.

### Cleanups
- Documentation updated along with types defined index.d.ts.
- Introduce [.npmignore](./.npmignore) which decreases package size from 145.3 kB to 10 kB

## v4.0.0
### BREAKING CHANGE
- Removed `userAgent` option in favor of `headers`. Please check [migration notes](./README.md#v4-migration).

## v3.2.0
### Documentation
- Added index.d.ts (thank you https://github.com/lil5)

## v3.1.0
### Changed
- Updated dependencies.

## v3.0.0
### Breaking changes
- A node-libcurl package is no longer being used in favor of `got`. This might lead to slightly different performance.
- Removed following properties from result: `nameLookupTime`, `connectTime`, `preTransferTime`, `startTransferTime`, `response`, `headerSize`, `header`, `headersParsed`
- Added `headers` property to result which now contains parsed headers object.

## v2.0.1
### Dependency updates
- node-libcurl update to v.2.1.1+

## v2.0.0
### Dependency updates (breaking changes)
- node-libcurl is updated to v2.1+ which contains some breaking changes e.x. dropped support for Node.js 4 and 6

## v1.2.2
### Dependency updates
- updated `node-libcurl` to 1.2.0 with exact version since newer has no release builds yet which makes the installation longer


test.js
const check = require('./index');

check({
  url: 'https://expired.badssl.com',
  keyword: 'agentslugdsfdsfsfsdfd',
  redirectsLimit: 2
})
  .then(res => {
    console.log(res);
  })
  .catch(err => {
    console.log(err);
  });


test/index-test.js
const chai = require('chai');
const assert = require('assert')

const check = require('../index');

chai.should();
const correctUrl = 'https://agentslug.com/en';
const redirectUrl = 'http://agentslug.com';
const notFoundUrl = 'https://agentslug.com/404-link';
const invalidUrl = 'invalidurl';

const propertiesRequired = [
  'httpCode',
  'httpCodeLang',
  'bodySize',
  'headers',
  'body',
  'status',
  'redirectsCount',
  'effectiveUrl',
  'totalTime'
];

describe('index', () => {
  describe('Uptime test', function () {
    it(`should test ${correctUrl} and finish with status: true`, function (done) {
      this.timeout(4000);
      check({
        url: correctUrl,
        keyword: 'agentslug'
      })
        .then((result) => {
          propertiesRequired.forEach(property => {
            result.should.have.property(property);
          });
          result.httpCode.should.be.a('number');
          result.httpCodeLang.should.be.a('string');
          result.status.should.be.equal(true);
          result.headers.should.be.an('object');
          // non regression - not full body is gathered
          result.body.length.should.be.equal(result.bodySize);

          return done();
        })
        .catch((err) => {
          setTimeout(() => {
            assert(false, err.message);
            return done();
          });
        })
    });
    it(`test ${correctUrl} without keyword and finish with status: true`, function (done) {
      this.timeout(4000);
      check({
        url: correctUrl
      })
        .then(result => {
          propertiesRequired.forEach(property => {
            result.should.have.property(property);
          });
          result.httpCode.should.be.a('number');
          result.httpCodeLang.should.be.a('string');
          result.status.should.be.equal(true);
          result.headers.should.be.an('object');

          return done();
        })
        .catch((err) => {
          setTimeout(() => {
            assert(false, err.message);
            return done();
          })
        })
    });
    it(`should test ${redirectUrl} with default redirects limit and finish with status: true`, function (done) {
      this.timeout(4000);
      check({
        url: redirectUrl
      })
        .then(result => {
          propertiesRequired.forEach(property => {
            result.should.have.property(property);
          });
          result.httpCode.should.be.a('number');
          result.httpCodeLang.should.be.a('string');
          result.status.should.be.equal(true);
          result.effectiveUrl.should.be.equal(correctUrl);
          result.redirectsCount.should.be.equal(2);
          result.headers.should.be.an('object');

          return done();
        })
        .catch((err) => {
          setTimeout(() => {
            assert(false, err.message);
            return done();
          })
        })
    });
    it(`should test ${redirectUrl} with no redirects limit and finish with status: false`, function (done) {
      this.timeout(4000);
      check({
        url: redirectUrl,
        redirectsLimit: 0,
      })
        .then(result => {
          propertiesRequired.forEach(property => {
            result.should.have.property(property);
          });
          result.httpCode.should.be.a('number');
          result.httpCodeLang.should.be.a('string');
          result.status.should.be.equal(false);
          result.headers.should.be.an('object');
          return done();
        })
    });
  });

  describe('Downtime test', function () {
    it(`should test ${notFoundUrl} and fail`, function (done) {
      this.timeout(4000);
      check({
        url: notFoundUrl,
      })
        .then(result => {
          result.status.should.be.equal(false);
          done();
        })
    });
  });

  describe('Error handling', () => {
    it('should reject on invalid url', () => {
      assert.rejects(() => {
        return check({
          url: invalidUrl,
        });
      });
    });
    it('should reject on missing url', () => {
      assert.rejects(() => {
        return check({});
      });
    });

    it('should reject on wrong type', () => {
      assert.rejects(() => {
        return check({
          url: 'https://example.com',
          redirectsLimit: 'yes'
        });
      });

      assert.rejects(() => {
        return check({
          url: 'https://example.com',
          timeOut: 'yes'
        });
      });

      assert.rejects(() => {
        return check({
          url: 'https://example.com',
          keyword: true
        });
      });

      assert.rejects(() => {
        return check({
          url: true,
        });
      });
    });
  });
});


__tests__/index.test.js
const check = require('../index');

const correctUrl = 'https://agentslug.com/en';
const redirectUrl = 'http://agentslug.com';
const notFoundUrl = 'https://agentslug.com/404-link';
const invalidUrl = 'invalidurl';

const propertiesRequired = [
  'httpCode',
  'httpCodeLang',
  'bodySize',
  'headers',
  'body',
  'status',
  'redirectsCount',
  'effectiveUrl',
  'totalTime'
];

describe('index', () => {
  const assertStandard = (result, status) => {
    propertiesRequired.forEach(property => {
      expect(typeof result[property]).not.toBe('undefined');
    });
    expect(typeof result.httpCode).toBe('number');
    expect(typeof result.bodySize).toBe('number');
    expect(typeof result.totalTime).toBe('number');
    expect(typeof result.httpCodeLang).toBe('string');
    expect(typeof result.headers).toBe('object');
    // non regression - not full body is gathered
    expect(result.body.length).toBe(result.bodySize);

    expect(result.status).toBe(status);
  };

  describe('Uptime test', () => {
    it(`should test ${correctUrl} and finish with status: true`, async () => {
      const result = await check({
        url: correctUrl,
        keyword: 'agentslug'
      })

      assertStandard(result, true);
    });

    it(`test ${correctUrl} without keyword and finish with status: true`, async () => {
      const result = await check({
        url: correctUrl
      });

      assertStandard(result, true);
    });

    it(`should test ${redirectUrl} with default redirects limit and finish with status: true`, async () => {
      const result = await check({
        url: redirectUrl
      });
      assertStandard(result, true);
    });

    it(`should test ${redirectUrl} with no redirects limit and finish with status: false`, async () => {
      const result = await check({
        url: redirectUrl,
        redirectsLimit: 0,
      });

      assertStandard(result, false);
    });
  });

  describe('Downtime test', () => {
    it(`should test ${notFoundUrl} and fail`, async () => {
      const result = await check({
        url: notFoundUrl,
      });

      assertStandard(result, false);
    });
  });

  describe('Bad SSL', () => {
    it('should fail', async () => {
      const result = await check({
        url: 'https://expired.badssl.com',
      });

      assertStandard(result, false);
      expect(result.errorCode).toBe('CERT_HAS_EXPIRED');
    });
  })

  describe('Error handling', () => {
    it('should reject on invalid url', () => {
      expect(() => {
        return check({
          url: invalidUrl,
        });
      }).rejects.toThrow();
    });
    it('should reject on missing url', () => {
      expect(() => {
        return check({});
      }).rejects.toThrow();
    });

    it('should reject on wrong type', () => {
      expect(() => {
        return check({
          url: 'https://example.com',
          redirectsLimit: 'yes'
        });
      }).rejects.toThrow();

      expect(() => {
        return check({
          url: 'https://example.com',
          timeOut: 'yes'
        });
      }).rejects.toThrow();

      expect(() => {
        return check({
          url: 'https://example.com',
          keyword: true
        });
      }).rejects.toThrow();

      expect(() => {
        return check({
          url: true,
        });
      }).rejects.toThrow();

      expect(() => {
        return check({
          url: 'https://example.com',
          headers: true,
        })
      }).rejects.toThrow();
    });
  });
});


index.d.ts
declare module "uptime-check" {
	export interface UptimeCheckResult {
    status: boolean; // true if test passed, false if failed
    httpCode: number;
		httpCodeLang: string; // English translation for http code (e.x. "Bad Request")
    redirectsCount: number,
		effectiveUrl: string; // last url in the redirects chain
		response: string; // full response (header + body)
		body: string; // body (response without header)
		bodySize: number; // body siz
		headers: Record<string, string>; // parsed headers
		totalTime: number; // total request time
		errorCode?: string; // in case of a request error, e.g. ETIMEDOUT, CERT_HAS_EXPIRED
	}

	export interface UptimeCheckOptions {
		url: string;
		keyword?: string; // defaults to null
		redirectsLimit?: number; // defaults to 3
		headers?: object;
		timeOut?: string; // defaults to 10 seconds
	}

	export default function check(o: UptimeCheckOptions): Promise<UptimeCheckResult>;
}


tasks/request.js
const got = require('got');
const HTTPCodesToLang = require('../lang/http-codes-to-lang');

const makeOptions = (options) => ({
  method: 'get',
  url: options.url,
  followRedirect: options.redirectsLimit > 0,
  maxRedirects: options.redirectsLimit,
  timeout: options.timeOut * 1000,
  headers: options.headers
});

const makeResponse = async (options) => {
  try {
    return await got(makeOptions(options))
  } catch (error) {
    if (error.response) {
      return error.response;
    }

    if (error.code) {
      return {
        statusCode: 0,
        url: options.url,
        request: {
          redirects: [],
        },
        headers: {},
        body: '',
        bodySize: 0,
        errorCode: error.code,
      }
    }

    throw error;
  }
};

/**
 * Makes curl request and prepares
 * result data.
 *
 * @param options
 * @return {Promise}
 */
const makeRequest = async (options) => {
  const now = Date.now();
  const codesToLang = new HTTPCodesToLang(options.timeOut);
  const response = await makeResponse(options);
  const totalTime = (Date.now() - now) / 1000;

  const result = {
    httpCode: response.statusCode,
    httpCodeLang: codesToLang.toLang(response.statusCode),
    totalTime,
    effectiveUrl: response.url,
    redirectsCount: response.request.redirects.length,
    headers: response.headers,
    body: response.body,
    bodySize: response.body ? response.body.length : 0,
    errorCode: response.errorCode,
  };

  return result;
}

module.exports = makeRequest;


lang/http-codes-to-lang.js
class HTTPCodesToLang {
  constructor(timeout) {
    this.timeout = timeout;
  }
  toLang(code) {
    if (!this.definitions.hasOwnProperty(code)) {
      return 'Not specified HTTP status';
    }
    return this.definitions[code];
  }
  get definitions() {
    return {
      0: `Not available or ${this.timeout}s timeout`,
      100: 'Continue',
      101: 'Switching Protocols',
      102: 'Processing',
      200: 'OK',
      201: 'Created',
      202: 'Accepted',
      203: 'Non-Authoritative Information',
      204: 'No Content',
      205: 'Reset Content',
      206: 'Partial Content',
      207: 'Multi-Status',
      300: 'Multiple Choices',
      301: 'Moved Permanently',
      302: 'Found',
      303: 'See Other',
      304: 'Not Modified',
      305: 'Use Proxy',
      306: 'Switch Proxy',
      307: 'Temporary Redirect',
      400: 'Bad Request',
      401: 'Unauthorized',
      402: 'Payment Required',
      403: 'Forbidden',
      404: 'Not Found',
      405: 'Method Not Allowed',
      406: 'Not Acceptable',
      407: 'Proxy Authentication Required',
      408: 'Request Timeout',
      409: 'Conflict',
      410: 'Gone',
      411: 'Length Required',
      412: 'Precondition Failed',
      413: 'Request Entity Too Large',
      414: 'Request-URI Too Long',
      415: 'Unsupported Media Type',
      416: 'Requested Range Not Satisfiable',
      417: 'Expectation Failed',
      418: 'I\'m a teapot',
      422: 'Unprocessable Entity',
      423: 'Locked',
      424: 'Failed Dependency',
      425: 'Unordered Collection',
      426: 'Upgrade Required',
      449: 'Retry With',
      450: 'Blocked by Windows Parental Controls',
      500: 'Internal Server Error',
      501: 'Not Implemented',
      502: 'Bad Gateway',
      503: 'Service Unavailable',
      504: 'Gateway Timeout',
      505: 'HTTP Version Not Supported',
      506: 'Variant Also Negotiates',
      507: 'Insufficient Storage',
      509: 'Bandwidth Limit Exceeded',
      510: 'Not Extended',
      520: 'Proxy publisher failure',
      521: 'Proxy publisher failure',
      522: 'Proxy publisher failure',
      524: 'Proxy publisher failure',
      523: 'Proxy publisher failure',
    };
  }
}

module.exports = HTTPCodesToLang;


index.js
const request = require('./tasks/request');

/**
 * Validates options. Throws Error if something is wrong.
 *
 * @param {UptimeCheckOptions} opts
 * @throws Error
 */
const validateOptions = (opts) => {
  const keysRequired = [
    'url'
  ];
  const numbers = [
    'redirectsLimit',
    'timeOut'
  ];
  const strings = [
    'keyword',
    'url'
  ];
  const objects = [
    'headers'
  ];

  keysRequired.forEach(key => {
    if (!opts.hasOwnProperty(key)) {
      throw new Error(`Missing required option: ${key}`);
    }
  });
  numbers.forEach(key => {
    if (typeof opts[key] !== 'number' || parseInt(opts[key]) !== opts[key]) {
      throw new Error(`Invalid option: ${key} must be an integer number`);
    }
  });

  strings.forEach(key => {
    if (opts[key] && typeof opts[key] !== 'string') {
      throw new Error(`Invalid option: ${key} must be a string.`);
    }
  });

  objects.forEach(key => {
    if (opts[key] && typeof opts[key] !== 'object') {
      throw new Error(`Invalid option: ${key} must be an object.`);
    }
  });

  // Only http and https can be tested.
  if (opts.url.startsWith('http') === false) {
    throw new Error('Url must be http:// or https://');
  }
}

/**
 * Merges options with defaults.
 * @param {UptimeCheckOptions} opts
 * @returns {UptimeCheckOptions}
 */
const mergeDefaults = (opts) => {
  const defaults = {
    keyword: null,
    redirectsLimit: 3,
    headers: {
      'User-Agent': 'Uptime-check - https://www.npmjs.com/package/uptime-check',
    },
    timeOut: 10
  };

  return Object.assign({}, defaults, opts);
}

/**
 *
 * @param {UptimeCheckOptions} opts
 * @returns {Promise<UptimeCheckResult>}
 */
const check = async (opts) => {
  /**
   *
   * @type {UptimeCheckOptions}
   */
  const options = mergeDefaults(opts);
  validateOptions(options)

  const result = await request(options);

  const httpCode = result.httpCode;

  result.status = (httpCode >= 200 && httpCode < 300);
  if (result.status && options.keyword) {
    const body = typeof result.body === 'undefined' ? '' : result.body;
    result.status = body.indexOf(options.keyword) > -1;
  }

  return result;
}


module.exports = check;
